docs/design/l2-peer-protocol-spec.md

```diff
@@ -0,0 +1,507 @@
+should be derived and its structure.
+
+> Key points:
+>
+> - we have a few parts of that phase: optional term discovery, channel opening
+>   with its setup negotiation and the final staging of the channel on-chain.
+> - up until the beginning of settlement both partners on disconnection MUST
+>   drop the channel.
+> - every error drops the channel and closes that message thread.
+
+#### The Full Conversation
+
+```mermaid
+ sequenceDiagram
+  participant opener
+  participant non-opener
+
+  opt Terms Discovery
+    opener ->> non-opener: QueryTerms
+    non-opener ->> opener: Terms | QueryError
+  end
+  opener ->> non-opener: Open | Error
+  non-opener ->> opener: Accept | Error
+  opener ->> non-opener: Staging
+  opener ->> non-opener: ChannelStaged | Error
+  non-opener ->> opener: ChannelStaged | Error
+```
+
+#### Terms Discovery
+
+To simplify the opening process there is an optional terms discovery. Accepted
+parameter ranges can change over time so they should be rediscovered during
+every new opening. We propose this mechanism to simplify the error reporting and
+handling.
+
+```haskell
+-- An inclusive range
+data Range a
+  = Range
+    { minValue :: a
+    , maxValue :: a
+    }
+
+data QueryTerms
+  = QueryTerms
+    { currency :: Currency
+    }
+
+data Terms
+  = Terms
+    { currency :: Currency
+    , htlcCountRange :: Range Integer
+    , htlcAmountRange :: Range Amount
+    , totalHtlcAmountRange :: Range Amount
+    , minGiftAmount :: Amount                -- Minimum "service fee" required
+    , minGiftPercent :: Integer              -- Used above the `minGiftAmount`
+    , peerFundingRange :: Range Amount       -- Minimum funding required from peer
+    , respondPeriodRange :: Range Milliseconds
+    , termsAndConditions :: Maybe Text
+    }
+
+data QueryError
+  = QueryError
+      { supportedCurrencies :: [(Currency, Text)]
+      }
+```
+
+This mechanism have obvious limitations:
+
+- Provided ranges can have internal dependencies - for example `responsePeriod`
+  can depend on the funds locked in a channel so even though the `opener` can
+  provide the final channel proposition which fullfiles per field requirements
+  it can still be rejected by the counterparty. More involved policies can use
+  `termsAndConditions` field to explain the details for the user.
+- The list is not exhaustive - it can be extended in the future.
+
+#### Opening and Staging
+
+> QUESTION:
+>
+> - Shoudl we expand the failure scenarios - transaction can timeout etc.?
+> - Should we discuss the transaction validity interval and in the future fee
+>   bumping etc.?
+
+The opening consists of a simple negotiation where the `openner` proposes the
+channel parameters and the `non-opener` accepts them with possible adjustments.
+Then the `opener` finally confirms the channel by sending a signed transaction
+which can be submitted by both parties. After this step both parties should
```

Comment:
submitted by both parties? 

Surely the opener just submits it, and awaits engagement from the other partner